docs/proposals/19-migrating-to-prometheus-operator.md.md

```diff
@@ -0,0 +1,444 @@
+          namespace: shoot--project--name
+        spec:
+          endpoints:
+          - port: metrics # scrape the service port with name `metrics`
+            bearerTokenFile: /var/run/secrets/gardener.cloud/shoot/token/token # could also be replaced with a secret
+            metricRelabelings:
+            - sourceLabels: [ __name__ ]
+              regex: ^(rest_client_requests_total|process_max_fds|process_open_fds)$
+              action: keep
+          namespaceSelector:
+            matchNames:
+            - shoot--project--name
+          selector:
+            matchLabels:
+              role: cloud-controller-manager # discover any service with this label
+        ```
+
+      ```yaml
+      apiVersion: monitoring.coreos.com/v1
+      kind: PrometheusRule
+      metadata:
+        labels:
+          cluster: shoot--project--name
+        name: cloud-controller-manager-rules
+        namespace: shoot--project--name
+      spec:
+        groups:
+        - name: cloud-controller-manager.rules
+          rules:
+          - alert: CloudControllerManagerDown
+            expr: absent(up{job="cloud-controller-manager"} == 1)
+            for: 15m
+            labels:
+              service: cloud-controller-manager
+              severity: critical
+              type: seed
+              visibility: all
+            annotations:
+              description: All infrastruture specific operations cannot be completed (e.g. creating loadbalancers or persistent volumes).
+              summary: Cloud controller manager is down.
+      ```
+
+    - Adding meta monitoring for the extensions: If the extensions need to be
+      scraped for monitoring, the extensions must bring their own [Custom
+      Resources][prom-crds].
+        - Currently the contract between extensions and gardener is that
+          anything that needs to be scraped must have the labels:
+          `prometheus.io/scrape=true` and `prometheus.io/port=<port>`. This is
+          defined [here][prom-config]. This is something that we can still
+          support with a `PodMonitor` that will scrape any pod in a specified
+          namespace with these labels.
+
+5. Add a [sidecar][grafana-sidecar] to grafana that will pickup dashboards and
+   provision them. Each dashboard gets its own configmap.
+
+   - Some of the dashboards were initially inspired by [kubernetes-mixins].
+     Gardener should reuse dashboards and alerting configs whenever possible. It
+     is not clear yet how to integrate this into Gardener.
+
+6. Migrating to the new monitoring stack:
+    1. Deploy the [prometheus-operator] and its custom resources.
+    1. Add a feature gate to optionally disable the current monitoring stack.
+    1. Co-deploy both monitoring stacks.
+    1. Ensure that the new stack provides similar coverage.
+    1. Switch over to the monitoring model defined in this GEP.
+    1. Deprecate the current monitoring stack.
+    1. Delete the current monitoring stack.
```

Comment:
What is your motivation behind the feature gate and the parallel deployment approach?

I think the parallel deployment approach will bring some drawbacks and confusion, e.g. we need two endpoints, etc.

We could also think about ensuring backwards-compatibility in the following way:
- keep collecting dashboards and scrape configs like today as defined by the current extension contract
- put all collected dashboards into a config map like today
- put all collected scrape config into a config map and use it as `additionalScrapeConfigs` in the `Prometheus` object

This way, we can ensure a smooth migration by giving extensions some time to adapt and we don't need a feature gate and an additional monitoring stack.
The compatibility-related code can be dropped after a while to complete the migration.

WDYT?